mypy/types.py

```diff
@@ -2177,9 +2177,18 @@ def function_type(func: mypy.nodes.FuncBase, fallback: Instance) -> FunctionLike
         return func.type
     else:
         # Implicit type signature with dynamic types.
-        # Overloaded functions always have a signature, so func must be an ordinary function.
-        assert isinstance(func, mypy.nodes.FuncItem), str(func)
-        return callable_type(func, fallback)
+        if isinstance(func, mypy.nodes.FuncItem):
+            return callable_type(func, fallback)
+        else:
+            # Broken overloads can have self.type set to None.
+            assert isinstance(func, mypy.nodes.OverloadedFuncDef)
```

Comment:
I wonder if it would be better to always set the overload type in the semantic analyzer. No need to do anything about this now, but consider adding a TODO comment.